packages/dds/tree/src/feature-libraries/chunked-forest/codec/schemaBasedEncoding.ts

```diff
@@ -64,12 +72,24 @@ export function fieldShaper(
 	treeHandler: TreeShaper,
 	field: TreeFieldStoredSchema,
 	cache: EncoderCache,
+	storedSchema: StoredSchemaCollection,
 ): FieldEncoder {
 	const kind = cache.fieldShapes.get(field.kind) ?? fail("missing FieldKind");
 	const type = oneFromSet(field.types);
 	const nodeEncoder = type !== undefined ? treeHandler.shapeFromTree(type) : anyNodeEncoder;
-	// eslint-disable-next-line unicorn/prefer-ternary
 	if (kind.multiplicity === Multiplicity.Single) {
+		if (field.kind === identifierFieldKindIdentifier) {
+			assert(type !== undefined, "field type must be defined in identifier field");
+			const nodeSchema = storedSchema.nodeSchema.get(type);
+			assert(nodeSchema !== undefined, "nodeSchema must be defined");
+			assert(
+				nodeSchema instanceof LeafNodeStoredSchema,
+				"nodeSchema must be LeafNodeStoredSchema",
+			);
+			assert(nodeSchema.leafValue === 1, "identifier field can only be type string");
+			const identifierNodeEncoder = new NodeShape(type, 0, [], undefined);
```

Comment:
can't remember if I asked this, but can `type` ever vary? or could we have a constant NodeShape object that we reuse?

Reply:
there is nothing at the encoding/stored schema layer that prevents type from varying. Nothing here takes a hard dependency on the exact schema used for identifier leave nodes (or any specific schema at all). Preserving this decoupling seem valuable. The alternative, taking a direct dependency on a specific schema, would both be a layering violation and a cyclic dep we would need to resolve somehow, and offer very little gain (note this is already cached by the EncoderCache, so its not like we are doing lots of allocations of this)